test/testdata/infer/glb_final.rb

```diff
@@ -0,0 +1,45 @@
+# typed: true
+extend T::Sig
+
+class FinalClass
+  extend T::Helpers
+  final!
+end
+module NonFinalModule; end
+
+class NonFinalClass; end
+module FinalModule
+  extend T::Helpers
+  final!
+end
+
+sig {params(x: FinalClass).void}
+def example1(x)
+  case x
+  when NonFinalModule
+    puts(x) # error: This code is unreachable
+  else
+    T.absurd(x) # error: Control flow could reach `T.absurd` because the type `FinalClass` wasn't handled
+  end
+end
+
+sig {params(x: NonFinalClass).void}
+def example2(x)
+  case x
+  when FinalModule
+    puts(x) # error: This code is unreachable
+  else
+    T.absurd(x) # error: Control flow could reach `T.absurd` because the type `NonFinalClass` wasn't handled
+  end
+end
+
+sig do
+  params(
+    x: T::Array[T.all(FinalClass, NonFinalModule)],
+    y: T::Array[T.all(NonFinalClass, FinalModule)]
+  ).void
+end
+def example3(x, y)
+  T.reveal_type(x) # error: `T::Array[T.noreturn]`
+  T.reveal_type(y) # error: `T::Array[T.noreturn]`
+end
```

Comment:
@froydnj When you say `NonFinalModuleIncludedInFinalClass` does that refer to a `module`?

I'm not sure how it would would make a difference whether that module was included into a final class at some later point. And if it's a class, it would be no different from the other `Class`/`Class` case (this PR only affects `Class`/`Module` and `Module`/`Class`)

Reply:
> @froydnj When you say `NonFinalModuleIncludedInFinalClass` does that refer to a `module`?
> 
> I'm not sure how it would would make a difference whether that module was included into a final class at some later point. And if it's a class, it would be no different from the other `Class`/`Class` case (this PR only affects `Class`/`Module` and `Module`/`Class`)

The sorbet.run link has an example of what I mean (the `module` is `include`'d in the definition of the `final` class, not at some later point).

I don't know that people would actually write out that kind of type by hand, but I can imagine it coming up in some kind of generic method context?

Reply:
ah sorry i was looking for `NonFinalModuleIncludedInFinalClass` in the snippet when i glanced and didn't realize that the concept was there but the name had been shortened.

nothing in the example looks to me like it would hit the logic in the glb case i've added here—it collapses because it hits the `derivesFrom` case earlier in the `if`/`else`, before ever looking at `isFinal`.

I think I'm going to hold off on adding a test for this because i don't want to wait for another CI run 😅